### PR TITLE
Implement `ChannelListing` for new chat design

### PR DESCRIPTION
**Pull request by jai-x**
Part of https://github.com/ppy/osu/issues/17001

Reference design: https://www.figma.com/file/f8b2dHp9LJCMOqYP4mdrPZ/Client%2FChat?node-id=2%3A63

https://user-images.githubusercontent.com/8362491/157338056-cfbb2079-04b0-4610-a335-17fceaee889a.mp4

Adds components `ChannelListing` and `ChannelListing` item with visual
test. Essentially a more simplified version of the existing
`ChannelSelectionOverlay` component.

Correctly implements `IFilterable` behaviour to filter child channel
items. Channel joined state is based on the underlying `Joined` bindable
of the `Channel` class.

Channel join/leave events are exposed via `OnRequestJoin` and
`OnRequestLeave` events which should be handled by parent component.

Requires a cached `OverlayColourScheme` instance to be provided by the
parent overlay component when added. Uses colour scheme `Pink` in
visual tests.

For now I have put the new components under the `NewChat` namespace
and plan to remove that namespace when properly integrating into the main
game. If this is not desired I can remove it.

### CODE REVIEW

**Comment by bdach** [comment]
Not sure about the namespacing choices here with `Online.NewChat` - generally there's no precedent for this, and what's the plan for old pieces? Is `NewChat` going to be there until the end of time?

Looks a bit funky in test browser, too:

2022-03-08-234630_248x69_scrot

File: osu.Game/Overlays/NewChat/ChannelListing.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+            foreach (var item in flow.Children)
+            {
+                item.OnRequestJoin += channel => OnRequestJoin?.Invoke(channel);
+                item.OnRequestLeave += channel => OnRequestLeave?.Invoke(channel);
+            }
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
```
Comment (bdach) [inline osu.Game/Overlays/NewChat/ChannelListing.cs:56]: BDL should be where the ctor is right now

**Comment by jai-x** [comment]
For now I have put the new components under the `NewChat` namespace and plan to remove that namespace when all the components have been added the overlay is properly integrated into the main game. I mainly did this just to make things clearer for me personally while developing. If this is not desired I can remove it.

File: osu.Game/Overlays/NewChat/ChannelListingItem.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                    channelText.Colour = normalColour;
+                }
+            }, true);
+
+            // Set action on click
+            Action = () => (channelJoined.Value ? OnRequestLeave : OnRequestJoin)?.Invoke(channel);
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
```
Comment (bdach) [inline osu.Game/Overlays/NewChat/ChannelListingItem.cs:101]: order all over the place here too - preferred ordering is:

* ctor
* BDL
* `LoadComplete()`
* event handlers/protected methods
* private methods

Also see: https://github.com/ppy/osu-framework/issues/2278

**Review by bdach — CHANGES_REQUESTED** [review]
aside from code quality inspections

File: osu.Game/Overlays/NewChat/ChannelListingItem.cs (hunk 2 above)
Comment (jai-x) [reply osu.Game/Overlays/NewChat/ChannelListingItem.cs]: I didn't know about this ordering guide. I will make sure to use it.

File: osu.Game/Overlays/NewChat/ChannelListingItem.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                                Margin = new MarginPadding { Bottom = 2 },
+                            },
+                            new OsuSpriteText
+                            {
+                                Anchor = Anchor.CentreLeft,
+                                Origin = Anchor.CentreLeft,
+                                Text = channel.Topic,
+                                Font = OsuFont.Torus.With(size: TEXT_SIZE),
+                                Margin = new MarginPadding { Bottom = 2 },
+                                Colour = Colour4.White,
```
[the remark this answers is not in the document]
Comment (jai-x) [reply osu.Game/Overlays/NewChat/ChannelListingItem.cs]: `Content1` seemed to resolve to white `RGB(1,1,1)` so I've used `Content2` instead.

File: osu.Game/Overlays/NewChat/ChannelListing.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                    Child = flow = new SearchContainer<ChannelListingItem>
+                    {
+                        Direction = FillDirection.Vertical,
+                        RelativeSizeAxes = Axes.X,
+                        AutoSizeAxes = Axes.Y,
+                        Spacing = new Vector2(3),
+                        Padding = new MarginPadding
+                        {
+                            Vertical = 13,
+                            Horizontal = 15,
```
[the remark this answers is not in the document]
Comment (jai-x) [reply osu.Game/Overlays/NewChat/ChannelListing.cs:58]: Would it be better to adjust sizing now or keep the sizing verbatim for now and make a post-pass after all the components have been added?

**Comment by peppy** [comment]
On the topic of namespace, do you actually have any conflicting class names? In this PR it looks like the naming you've chosen differs from existing classes, at which point I'd not create a new namespace at all if that's feasible.

Rationale is that following git changes across files is a bit of a pain so where possible we should try and put files in their final home from the outset.

File: osu.Game/Overlays/NewChat/ChannelListing.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Overlays/NewChat/ChannelListing.cs]: I personally do it as I open PRs for the individual pieces. A giant downsizing PR at the end of everything may not be super great, I dunno.

**Comment by peppy** [comment]
I've made a couple of minor tweaks, explained in the commit messages.

Looks good otherwise!

**Review by peppy — APPROVED** [review]

**Pull request merged**